staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -629,17 +629,22 @@ const (
 	// maxWatchChanSizeWithoutIndex is the max size of the channel
 	// used by the watch not using the index.
 	// TODO(wojtek-t): Figure out if the value shouldn't be higher.
-	maxWatchChanSizeWithoutIndex = 100
+	maxWatchChanSizeWithoutIndex = 1000
 )
 
-func (w *watchCache) suggestedWatchChannelSize(indexExists, triggerUsed bool) int {
+func (w *watchCache) suggestedWatchChannelSize(indexExists, triggerUsed bool, watcherCount int, terminatedWatchCount int) int {
 	// To estimate the channel size we use a heuristic that a channel
 	// should roughly be able to keep one second of history.
 	// We don't have an exact data, but given we store updates from
 	// the last <eventFreshDuration>, we approach it by dividing the
 	// capacity by the length of the history window.
 	chanSize := int(math.Ceil(float64(w.currentCapacity()) / eventFreshDuration.Seconds()))
 
+	// Increase chanSize by 10 for every 100 watchers
+	// Increase chanSize by 10 for every time that a watcher terminated due to unresponsiveness
+	chanSize += int(math.Ceil(float64(watcherCount)/100.0)) * 10
+	chanSize += terminatedWatchCount * 10
```

Comment:
This is very problematic.
This can be caused by slow client (e.g. we have a slow client that just doesn't read events and is periodically shutdown). If apiserver is running for long enough, terminatedWatchCount will grow to a large value and pretty much every watch will have a large channel.
This change isn't safe to do.